docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,844 @@
+    profile: aws
+    region: eu-west-1
+    secretBindingRef:
+      name: core-aws
+    providerConfig:
+      apiVersion: aws.cloud.gardener.cloud/v1alpha1
+      kind: InfrastructureConfig
+      networks:
+        vpc: # specify either 'id' or 'cidr'
+        # id: vpc-123456
+          cidr: 10.250.0.0/16
+        internal:
+        - 10.250.112.0/22
+        public:
+        - 10.250.96.0/22
+        workers:
+        - 10.250.0.0/19
+      zones:
+      - eu-west-1a
+    workerPools:
+    - name: pool-01
+      osType: CoreOS
+      autoScalerMin: 2
+      autoScalerMax: 2
+      providerConfig:
+        apiVersion: aws.cloud.gardener.cloud/v1alpha1
+        kind: WorkerPoolConfig
+        ami: ami-d0dcef3b
+        machineType: m4.large
+        volumeType: gp2
+        volumeSize: 20Gi
+        zones:
+        - eu-west-1a
+  kubernetes:
+    version: 1.11.0
+    ...
+  dns:
+    provider: aws-route53
+    domain: johndoe-aws.garden-dev.example.com
+  maintenance:
+    timeWindow:
+      begin: 220000+0100
+      end: 230000+0100
+    autoUpdate:
+      kubernetesVersion: true
+  backup:
+    schedule: "*/5 * * * *"
+    maximum: 7
+  addons:
+    kube2iam:
+      enabled: false
+    kubernetes-dashboard:
+      enabled: true
+    cluster-autoscaler:
+      enabled: true
+    nginx-ingress:
+      enabled: true
+      loadBalancerSourceRanges: []
+    kube-lego:
+      enabled: true
+      email: john.doe@example.com
+    monocular:
+      enabled: false
+```
+
+:information: The specifications for the other already existing cloud providers looks similar.
+
+### CRD definitions and workflow adaptation
+
+In the following we are outlying the CRD definitions which will be one part of the contract between Gardener and the dedicated controllers.
+After that we will take a look at the current [reconciliation](https://github.com/gardener/gardener/blob/master/pkg/controller/shoot/shoot_control_reconcile.go)/[deletion](https://github.com/gardener/gardener/blob/master/pkg/controller/shoot/shoot_control_delete.go) flow and describe how it would look like in case we would implement this proposal.
+
+#### Custom resource definitions
+
+Every CRD has a `.spec.type` field containing the respective instance of the dimension the CRD represents, e.g. the cloud provider, the DNS provider or the operation system name.
+Moreover, the `.status` field must contain
+
+* `observedGeneration` (`int`), a field indicating on which generation the controller last worked on.
+* `state` (`*runtime.RawExtension`), a field which is not interpreted by Gardener but persisted; it should be treated opaque and only be used by the respective CRD-specific controller (it can store anything it needs to re-construct its own state).
+* `lastError` (`object`), a field which is optional and only present if the last operation ended with an error state.
+* `lastOperation` (`object`), a field which always exists and which indicates what the last operation of the controller was.
```

Comment:
I didn't see any [event](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#event-v1-core) usage on gardener, i'm curious if there is a historical reason and decision behind.

But if not so, why not using events and instead of having these `last*` fields. I think the right thing to do here is to use events, since there are actual events and not status, these should go to the events and not to the status fields.

Reply:
Only the last operation is most relevant or "informative" for the user. Gardener already uses `Events` for `Shoot` resources (check `kubectl describe shoot`), and new controllers can do the same. We want to keep the `lastOperation` and `lastError` fields to make it easier to check the current status. Parsing/Finding events for a resource is possible but difficult.